### ISSUE TITLE
Watchtower keeps old default environment variables after updating

### ISSUE THREAD

**Issue by reconman** [containrrr/watchtower#1342]
**Describe the bug**

Many containers have default values for environment variables, which change for each image version. For example, pymedusa/medusa includes the commit hash and branch.

Updating the container via watchtower keeps the old commit hash value. This leads to some wrong information in the app itself.

I think the same is also true for labels.

**To Reproduce**
1. Build your own docker image with a default value for an environment variable
2. Create a docker container from that image
3. Build a new image version with the same tag, but change the default value of the environment variable
4. Run watchtower
5. Check the current environment variables of the updated container. The environment variable will have the old default value. 

**Expected behavior**

If a container is updated, any environment variable default values should be replaced with the new default values.

Maybe this can be achieved by checking the default environment variables of the old image and removing them from the new `docker run` command.

**Environment**
Platform: Synology
Architecture: amd64
Docker version: 20.10.3

**Comment by simonaronsson** [comment]
> **Describe the bug**
> 
> Many containers have default values for environment variables, which change for each image version. For example, pymedusa/medusa includes the commit hash and branch.
> 
> Updating the container via watchtower keeps the old commit hash value. This leads to some wrong information in the app itself.
> 
> I think the same is also true for labels.
> 
> **To Reproduce**
> 
> 1. Build your own docker image with a default value for an environment variable
> 2. Create a docker container from that image
> 3. Build a new image version with the same tag, but change the default value of the environment variable
> 4. Run watchtower
> 5. Check the current environment variables of the updated container. The environment variable will have the old default value.
> 
> **Expected behavior**
> 
> If a container is updated, any environment variable default values should be replaced with the new default values.
> 
> Maybe this can be achieved by checking the default environment variables of the old image and removing them from the new `docker run` command.
> 
> **Environment** Platform: Synology Architecture: amd64 Docker version: 20.10.3

I don't think this is achievable in a reliable way. How would watchtower know whether the value was put there intentionally or if it came from a default?

**Comment by reconman** [comment]
`docker history --format "{{.CreatedBy}}" --no-trunc pymedusa/medusa:develop | grep -o ENV.*`
results in
```
ENV MEDUSA_COMMIT_HASH=f778efb4a4d23a9964e66b681e8c4e2592288e3a
ENV MEDUSA_COMMIT_BRANCH=develop
ENV PYTHON_GET_PIP_SHA256=ba3ab8267d91fd41c58dbce08f76db99f747f716d85ce1865813842bb035524d
ENV PYTHON_GET_PIP_URL=https://github.com/pypa/get-pip/raw/6ce3639da143c5d79b44f94b04080abf2531fd6e/public/get-pip.py
ENV PYTHON_SETUPTOOLS_VERSION=58.1.0
ENV PYTHON_PIP_VERSION=22.0.4
ENV PYTHON_VERSION=3.10.4
ENV GPG_KEY=A035C8C19219BA821ECEA86B64E628F8D684696D
ENV LANG=C.UTF-8
ENV PATH=/usr/local/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
```

So it's definitely possible to get the ENV variables set in the Dockerfile.

**Comment by reconman** [comment]
`docker inspect --format "{{.Config.Env}}" pymedusa/medusa:develop` lists the default environment variables of the image.

`docker inspect --format "{{.Config.Env}}" medusa` lists the current environment variables of the container called "medusa".

Remove the first list from the second and only the environment variables set by the user are left.

Tried it in bash:
```bash
image_output=$(docker inspect --format "{{range .Config.Env}}{{println .}}{{end}}" pymedusa/medusa:develop)
container_output=$(docker inspect --format "{{range .Config.Env}}{{println .}}{{end}}" medusa)
echo ${image_output[@]} ${container_output[@]} | tr ' ' '\n' | sort | uniq -u
```
Result:
```
PGID=65537
PUID=1034
TZ=Europe/Berlin
```

That's exactly what I set.

**Comment by reconman** [comment]
I searched a bit and found that there's already code for subtracting the environment variables present: https://github.com/containrrr/watchtower/blob/main/pkg/container/container.go#L254

But it doesn't seem to work for me.

**Comment by simonaronsson** [comment]
> `docker inspect --format "{{.Config.Env}}" pymedusa/medusa:develop` lists the default environment variables of the image.
> 
> `docker inspect --format "{{.Config.Env}}" medusa` lists the current environment variables of the container called "medusa".
> 
> Remove the first list from the second and only the environment variables set by the user are left.
> 
> Tried it in bash:
> 
> ```shell
> image_output=$(docker inspect --format "{{range .Config.Env}}{{println .}}{{end}}" pymedusa/medusa:develop)
> container_output=$(docker inspect --format "{{range .Config.Env}}{{println .}}{{end}}" medusa)
> echo ${image_output[@]} ${container_output[@]} | tr ' ' '\n' | sort | uniq -u
> ```
> 
> Result:
> 
> ```
> PGID=65537
> PUID=1034
> TZ=Europe/Berlin
> ```
> 
> That's exactly what I set.

My question remains however, what happen if the user sets an environment variable to the very same value as the default, and then the image maintainer decides to change that default? Should the value be kept? Removed?

**Comment by reconman** [comment]
> My question remains however, what happen if the user sets an environment variable to the very same value as the default, and then the image maintainer decides to change that default? Should the value be kept? Removed? 

You can't detect that case, so you have to treat it as "the value was set by the image".

**Comment by simonaronsson** [comment]
> > My question remains however, what happen if the user sets an environment variable to the very same value as the default, and then the image maintainer decides to change that default? Should the value be kept? Removed?
> 
> You can't detect that case, so you have to treat it as "the value was set by the image".

Which is why I'm hesitant to add this feature. @piksel, what's your take on this?

**Comment by reconman** [comment]
Like I said, it's a bugfix for a broken feature, not a new one.

**Comment by Chri-s** [comment]
Just an idea: How about solving this with a label. For example `com.centurylinklabs.watchtower.env.keep.<env-name>=true` for environment variables that should not be updated?

**Comment by JohannesFleischer** [comment]
> Just an idea: How about solving this with a label. For example `com.centurylinklabs.watchtower.env.keep.<env-name>=true` for environment variables that should not be updated?

That would be great, because for example secrets that are hold solely as an env var in memory have to be kept because the container will not work without them.
But I think that a pattern would be even better so that if there are many secrets they can be named in a way that only one line is needed to keep them all.
